possibility to create htu with existing i2c object

kevinkk525 opened the pull request:
Adds the possibility to create a htu object with an existing i2c object, so you can create it before and use it afterwards without knowing which devices created the i2c object. 
I think this should be possible.
Many libraries just expect an i2c object, so now it is up to the user.

Also the temperature and humidity are functions now. Makes more sense to me especially as I was converting the library to be asynchronous (next commit).

What do you think?
Thanks for your work,

julianhille commented on htu21d.py:
```diff
@@ -43,30 +44,30 @@ def _crc_check(self, value):
             write_address (int): address to write to
         :return:
         """
+        import time
```
an import does not belong here.

kevinkk525 replied:
you are totally right.

julianhille commented on htu21d.py:
```diff
@@ -43,30 +44,30 @@ def _crc_check(self, value):
         self.i2c.writeto_mem(int(self.ADDRESS), int(write_address), '')
         self.i2c.stop()
         data = bytearray(3)
+        time.sleep_ms(50)
```
im not really sure if this is needed. i did not need it at all. Will see if `readfrom_into` is a blocking request.

kevinkk525 replied:
according to the datasheet (https://cdn-shop.adafruit.com/datasheets/1899_HTU21D.pdf) you should wait at least 50ms before reading the result. All arduino libraries do the same.
Can't remember if I added this because I had problems, but I probably did.

julianhille commented on htu21d_async.py:
```diff
@@ -0,0 +1,84 @@
+from machine import I2C, Pin
+import uasyncio as asyncio
+
+class HTU21D(object):
```
why do you copy the complete class instead of inherit it?

kevinkk525 replied:
I tested adding async functions to the original library first but my tests showed me that I could save RAM by copying the class and removing the sync functions, so I went for that way.

julianhille commented on htu21d_async.py:
```diff
@@ -0,0 +1,84 @@
+
+    async def humidity_async(self):
+        raw =  await self._issue_measurement_async(self.ISSUE_HU_ADDRESS)
+        if raw is None or raw==0:
```
you could also do a `if not raw` or is there any specific reason to do not?

julianhille replied:
and why is 0 a bad value? 0 is perfectly fine and would give a relative humidity value of -6% which is the minimum of the htu.

kevinkk525 replied:
0 theoretically is fine but sometimes I got the raw value 0 which was obviously an error so I tried to add the 0 to the errors.

julianhille commented on htu21d_async.py:
```diff
@@ -0,0 +1,84 @@
+            
+
+    async def temperature_async(self):
+        raw= await self._issue_measurement_async(self.ISSUE_TEMP_ADDRESS)
```
missing space before `=`

julianhille commented on htu21d_async.py:
```diff
@@ -0,0 +1,84 @@
+
+    async def temperature_async(self):
+        raw= await self._issue_measurement_async(self.ISSUE_TEMP_ADDRESS)
+        if raw is None or raw==0:
```
missing spaces around comparator

julianhille replied:
same here 0 is fine. and results to -46C

kevinkk525 replied:
0 theoretically is fine but sometimes I got the raw value 0 which was obviously an error so I tried to add the 0 to the errors.

julianhille commented on htu21d.py:
```diff
@@ -6,24 +6,25 @@ class HTU21D(object):
+            self.i2c=i2c
+        else:
+            self.i2c = I2C(scl=Pin(scl), sda=Pin(sda), freq=100000)
+        self.running=False
```
For what reason is the running member variable here?

kevinkk525 replied:
To be precise, it is of no importance here and can indeed be removed again.
This is a leftover of my port to async.

julianhille commented on htu21d.py:
```diff
@@ -6,24 +6,25 @@ class HTU21D(object):
 
     def _crc_check(self, value):
         """CRC check data
-
```
please don't remove empty lines here.

kevinkk525 replied:
sorry

julianhille commented on htu21d.py:
```diff
@@ -6,24 +6,25 @@ class HTU21D(object):
         
         Args:
             value (bytearray): data to be checked for validity
-
```
please don't remove empty lines here.

julianhille commented on htu21d.py:
```diff
@@ -43,30 +44,30 @@ def _crc_check(self, value):
 
     def _issue_measurement(self, write_address):
         """Issue a measurement.
-
```
please don't remove empty lines here.

julianhille commented on htu21d.py:
```diff
@@ -43,30 +44,30 @@ def _crc_check(self, value):
         raw &= 0xFFFC
-        return raw
-
-    @property
```
why do you issue an api change here and remove the property decorator?

kevinkk525 replied:
This is the first library for a temperature sensor that I have seen using a property for temperature and humidity. that alone was a reason for me to change the api, so it is similar to the other libraries. 
Also working towards an async api meant dropping the property as this is not supported with asyncio.

julianhille commented on htu21d_async.py:
```diff
@@ -0,0 +1,84 @@
+
+    def __init__(self, scl=None, sda=None, i2c=None):
+        """Initiate the HUT21D
+        Args:
```
missing new line

julianhille commented on htu21d_async.py:
```diff
@@ -0,0 +1,84 @@
+    async def _issue_measurement_async(self,write_address):
+        while self.running:
+            await asyncio.sleep_ms(160)
+        self.running=True
```
missing spaces around assignment

julianhille commented on htu21d_async.py:
```diff
@@ -0,0 +1,84 @@
+            self.i2c.stop()
+            data=bytearray(3)
+        except:
+            print("Error in HTU instruction")
```
print statements aren't good ideas for error reporting.
I would like to see an exception raised if something happens there instead of a print
and silently ignoring the error and just returning none.
I also would like to see a more explicit exception catch instead of a "catch" all.

kevinkk525 replied:
You are right, the print was a leftover from debugging, errors in these parts mostly indicate that the sensor is not connected (or not working correctly). Have to check the exact error to catch it. Did not pay attention to that as I only wanted to add the running = False which is basically a (cheap) lock preventing multiple readings at the same time, as this is possible with async. Or should that lock be removed and one should just handle this in code using the sensor?

julianhille commented on htu21d_async.py:
```diff
@@ -0,0 +1,84 @@
+            data=bytearray(3)
+        except:
+            print("Error in HTU instruction")
+            self.running=False
```
missing spaces around assignment

julianhille commented on htu21d_async.py:
```diff
@@ -0,0 +1,84 @@
+            self.i2c.readfrom_into(self.ADDRESS, data)
+            if not self._crc_check(data):
+                raise ValueError()
+            raw=(data[0]<<8)+data[1]
```
missing spaces around assignment.

julianhille commented:
Btw. im sorry that i forgot to thank you for the very first PR ;P

kevinkk525 commented:
Was using your library in my esp8266 smart home system and modified it for asyncio and wanted to share it back to you. Was my first pull request, so thanks for your detailed answers and especially for the library!

Please check the outdated ones too, github got some of those wrong.. they are still in the code, i just added the requested lines and spaces around assignments.
